h2o-persist-hdfs/src/main/java/water/persist/security/HdfsDelegationTokenRefresher.java

```diff
@@ -0,0 +1,215 @@
+
+    public static void setup(Configuration conf, String tmpDir) throws IOException {
+        boolean enabled = conf.getBoolean(H2O_AUTH_TOKEN_REFRESHER_ENABLED, false);
+        if (!enabled) {
+            log("HDFS Token renewal is not enabled in configuration", null);
+            return;
+        }
+        String authUser = conf.get(H2O_AUTH_USER);
+        String authPrincipal = conf.get(H2O_AUTH_PRINCIPAL);
+        String authKeytab = conf.get(H2O_AUTH_KEYTAB);
+        if (authKeytab == null) {
+            log("Keytab not provided, HDFS tokens will not be refreshed by H2O and their lifespan will be limited", null);
+            return;
+        }
+        double intervalRatio = Double.parseDouble(conf.get(H2O_AUTH_TOKEN_REFRESHER_INTERVAL_RATIO, "0.4"));
+        String authKeytabPath = writeKeytabToFile(authKeytab, tmpDir);
+        new HdfsDelegationTokenRefresher(authPrincipal, authKeytabPath, authUser, intervalRatio).start();
+    }
+
+    private static String writeKeytabToFile(String authKeytab, String tmpDir) throws IOException {
+        FileUtils.makeSureDirExists(tmpDir);
+        File keytabFile = new File(tmpDir, "hdfs_auth_keytab");
+        byte[] byteArr = BinaryFileTransfer.convertStringToByteArr(authKeytab);
+        BinaryFileTransfer.writeBinaryFile(keytabFile.getAbsolutePath(), byteArr);
+        return keytabFile.getAbsolutePath();
+    }
+
+    private final ScheduledExecutorService _executor = Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder().setDaemon(true).setNameFormat("hdfs-token-refresher-%d").build()
+    );
+
+    private final String _authPrincipal;
+    private final String _authKeytabPath;
+    private final String _authUser;
+    private final double _intervalRatio;
+
+    public HdfsDelegationTokenRefresher(
+            String authPrincipal,
+            String authKeytabPath,
+            String authUser,
+            double intervalRatio
+    ) {
+        _authPrincipal = authPrincipal;
+        _authKeytabPath = authKeytabPath;
+        _authUser = authUser;
+        _intervalRatio = intervalRatio;
+    }
+
+    public void start() {
+        long interval;
+        try {
+             interval = getTokenRenewalInterval(loginAuthUser());
+        } catch (IOException | InterruptedException e) {
+            log("Failed to determine delegation token renewal interval", e);
+            return;
+        }
+        final long actualInterval;
+        if (interval == Long.MAX_VALUE) {
+            log("Couldn't determine tokens renewal interval, will use 12h", null);
+            actualInterval = 12 * 3600 * 1000; // 12h
+        } else
+            actualInterval = (long) (interval * _intervalRatio);
+        log("Determined token renewal interval = " + interval + "ms. Using actual interval = " + actualInterval + "ms.", null);
+        _executor.scheduleAtFixedRate(this, 0, interval, TimeUnit.MILLISECONDS);
```

Comment:
nit: It seems to me that milliseconds don't bring much benefits in this logic. Maybe seconds could play the same role and not to generate so big numbers in logs.

Reply:
good point! will look into it

Reply:
fixed, thank you for the suggestion